### DISCUSSION TITLE
RangeError: Property storage exceeds 196607 properties in offline exchange

### DISCUSSION THREAD

**Discussion by Hassaan68** [urql-graphql/urql#2969] [Q&A]
hi mates, I am getting this exception in urql offline exchange storage in react native, The error is caused by makeAsyncStorage method of @urql/storage-rn because query returns the large amount of data. how can I resolve this issue?? as I want offline support in my app??

**Comment by kitten** [comment]
It'd be helpful if you could post a full stack and message in the meantime ✌️ I don't recognise this right off the bat, so I'd have to look up what this is related to

**Reply by JoviDeCroock** [reply]
Looks to be related to hermes segmenting the memory https://github.com/facebook/hermes/issues/139#issuecomment-543810007 so maybe an enormous payload gets normalized in way too many keys

**Reply by kitten** [reply]
To be fair I actually realised that this number is a constant limit. Basically, Hermes seems to have a hard coded limit for how large JS dictionaries may grow https://github.com/facebook/hermes/issues/851

cc @kadikraman 

I don't remember right away if the RN storage includes chunking. Because if it does, there's potentially an easy-ish fix of just adding a property limit for chunking in JS 🤔

**Reply by Hassaan68** [reply]
> It'd be helpful if you could post a full stack and message in the meantime ✌️ I don't recognise this right off the bat, so I'd have to look up what this is related to
![Screenshot_20230221_105929](https://user-images.githubusercontent.com/35756574/220331693-2dbb430e-a4a4-479a-bccb-92bb784e2c4d.jpg)

@kitten , it occurs at multiple location but I am attaching the one of the screenshots

**Reply by Hassaan68** [reply]
@kitten any way of excluding one query field for cache?

**Comment by kitten** [comment]
@Hassaan68 You can write a wrapper around one of the storage interfaces. Basically, the interface can be filtered as you need: https://github.com/urql-graphql/urql/blob/3fe3cd140ec5514beb27aaf4f46837126f7f2e6d/exchanges/graphcache/src/types.ts#L229-L235

This won't resolve your issue however. If you're letting a batch of data write >196,607 cache entries in one go, then internally, Graphcache will still fail because we didn't expect that any JS engine would fail on creating a large JS dictionary. So, we'll have to fix that either way.

Also on a side note, please be patient and don't create duplicate discussions just yet for what's essentially an extension to the same topic. We'll get to replying to you but it won't always be within an hour.

**Reply by Hassaan68** [reply]
sure @kitten , as this issue is critical so I am expecting the fix soon. and just let you know that Cache making app performance very slow specially in case of large amount of data

**Reply by JoviDeCroock** [reply]
@Hassaan68 that's a fix you can use, we're still OSS maintainers not people out there to fix everyone's stuff 😅 PR's are always welcome. GraphCache according to our benchmarks is pretty fast compared to other solutions but if you think other solutions are better go for that
